serial-settings/src/lib.rs

```diff
@@ -0,0 +1,374 @@
+//! Persistent Settings Management Serial Interface
+//!
+//! # Description
+//! This crate provides a simple means to load, configure, and save device settings over a serial
+//! (i.e. text-based) interface. It is ideal to be used with serial ports and terminal emulators,
+//! and exposes a simple way to allow users to configure device operation.
+//!
+//! # Example
+//! Let's assume that your settings structure looks as follows:
+//! ```rust
+//! #[derive(miniconf::Tree, ...)]
+//! struct Settings {
+//!     broker: String,
+//!     id: String,
+//! }
+//! ```
+//!
+//! A user would be displayed the following terminal interface:
+//! ```
+//!> help
+//! AVAILABLE ITEMS:
+//!   list
+//!   get <item>
+//!   set <item> <value>
+//!   clear
+//!   platform <cmd>
+//!   help [ <command> ]
+//!
+//! > plaform dfu
+//! Reset to DFU is not supported
+//!
+//! > plaform service
+//! Service data not available
+//!
+//! > list
+//! Available settings:
+//! /broker: "test" [default: "mqtt"]
+//! /id: "04-91-62-d2-a8-6f" [default: "04-91-62-d2-a8-6f"]
+//! ```
+//!
+//! # Design
+//! Settings are specified in a [`Miniconf::Tree`] settings tree and are transferred over the
+//! serial interface using JSON encoding. This means that things like strings must be encased in
+//! qutoes.
+//!
+//! # Limitations
+//! Currently, there is a hardcoded limit of 32-bytes on the settings path. This is arbitrary and
+//! can be changed if needed.
+#![no_std]
+
+use core::fmt::Write;
+use embedded_io::{Read, ReadReady, Write as EioWrite, WriteReady};
+use miniconf::{JsonCoreSlash, TreeKey};
+
+/// Specifies the API required for objects that are used as settings with the serial terminal
+/// interface.
+pub trait Settings: for<'a> JsonCoreSlash<'a> + Clone {
+    /// Reset the settings to their default values.
+    fn reset(&mut self) {}
+}
+
+pub trait Platform: Sized {
+    /// This type specifies the interface to the user, for example, a USB CDC-ACM serial port.
+    type Interface: embedded_io::Read
+        + embedded_io::ReadReady
+        + embedded_io::Write
+        + embedded_io::WriteReady;
+
+    /// Specifies the settings that are used on the device.
+    type Settings: Settings;
+
+    /// `load()`/`save()` Error type
+    type Error: core::fmt::Debug;
+
+    /// Load the settings from storage
+    fn load(&mut self, buffer: &mut [u8]) -> Result<(), Self::Error>;
+
+    /// Save the settings to storage
+    fn save(&mut self, buffer: &mut [u8]) -> Result<(), Self::Error>;
+
+    /// Execute a platform specific command.
+    fn cmd(&mut self, cmd: &str);
+
+    /// Return a mutable reference to the `Interface`.
+    fn interface_mut(&mut self) -> &mut Self::Interface;
+
+    /// Return a reference to the `Settings`
+    fn settings(&self) -> &Self::Settings;
+
+    /// Return a mutable reference to the `Settings`.
+    fn settings_mut(&mut self) -> &mut Self::Settings;
```

Comment:
I'm not following why `P::Settings` aren't just a member of `Context`. It seems like extra work for the implementor of `platform` to have to provide these when in reality the settings are internally managed by this library.

Reply:
To me "proper ownership" doesn't have clear and easy rules.

Settings changing outside this crate is something I can see coming up soon, e.g. also saving the application settings in flash and using that as the startup value (shadowed then later by mqtt where available and retained, I think I mentioned this somewhere else as well).